packages/aws_common/lib/src/http/aws_http_client.dart

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:aws_common/aws_common.dart';
+import 'package:aws_common/src/http/aws_http_client_io.dart'
+    if (dart.library.js) 'package:aws_common/src/http/aws_http_client_js.dart';
+import 'package:meta/meta.dart';
+
+/// {@template aws_common.http.aws_http_client}
+/// An HTTP client with support for HTTP/1.1, HTTP/2, and cancelable requests.
+/// {@endtemplate}
+abstract class AWSHttpClient implements Closeable {
+  /// {@macro aws_common.http.aws_http_client}
+  factory AWSHttpClient() = AWSHttpClientImpl;
+
+  /// Subclass constructor for [AWSHttpClient].
+  ///
+  /// **Internal use only**. For external cases, consider subclassing
+  /// [AWSBaseHttpClient] or [AWSCustomHttpClient] instead.
+  @protected
+  @internal
+  AWSHttpClient.protected();
+
+  /// Callback for VM clients when an SSL exception occurs due to an untrusted
+  /// or unverifiable certificate.
+  ///
+  /// Clients can support establishing an SSL connection despite this exception
+  /// by selectively returning `true` for these options. For example, when
+  /// debugging with local, self-signed certificates, it can be helpful to
+  /// return `true` from this method for all certificates.
+  BadCertificateCallback onBadCertificate = (_, __, ___) => false;
+
+  /// The supported HTTP protocols, used for negotiating with remote servers.
+  ///
+  /// By default, only HTTP/2 and HTTP/3 servers are supported.
+  SupportedProtocols supportedProtocols = SupportedProtocols.http2_3;
+
+  /// Sends [request] using the underlying HTTP protocol and returns the
+  /// streaming response.
+  AWSHttpOperation send(
+    AWSBaseHttpRequest request, {
+    FutureOr<void> Function()? onCancel,
+  });
+
+  @override
+  FutureOr<void> close({bool force = false});
+}
+
+/// Base class for all implementations of [AWSHttpClient].
+///
+/// This class can be used for customizing how requests are sent. For simple
+/// modifications in the request or response objects, like adding headers, it
+/// is easier to subclass [AWSBaseHttpClient].
+abstract class AWSCustomHttpClient extends AWSHttpClientImpl {}
```

Comment:
Yeah, actually the idea is to incentivize AWSBaseHttpClient for most users since it involves no handling of `CancelableOperation` which are pretty messy to work with.